session/session.go

```diff
@@ -540,7 +541,31 @@ func (s *session) CommitTxn(ctx context.Context) error {
 
 	var commitDetail *execdetails.CommitDetails
 	ctx = context.WithValue(ctx, execdetails.CommitDetailCtxKey, &commitDetail)
-	err := s.doCommitWithRetry(ctx)
+	var err error
+	txnScope := s.GetSessionVars().TxnCtx.TxnScope
+	if txnScope != config.DefTxnScope {
+		is := infoschema.GetInfoSchema(s)
+		for physicalTableID := range s.GetSessionVars().TxnCtx.TableDeltaMap {
+			bundle, ok := is.BundleByName(placement.GroupID(physicalTableID))
+			if !ok {
+				err = fmt.Errorf("table or partition %v don't have rule bundle with non-global txnScope %v changing", physicalTableID, txnScope)
+				break
+			}
+			dcLocation, ok := placement.GetLeaderDCByBundle(bundle, placement.DCLabelKey)
+			if !ok {
+				err = fmt.Errorf("bundle %v 's leader placement policy is not clear", bundle.ID)
+				break
+			}
+			if dcLocation != txnScope {
+				err = fmt.Errorf("table or partition %v's leader location %v is out of txn_scope %v", physicalTableID, dcLocation, txnScope)
+				break
+			}
+		}
+		if err != nil {
+			goto EXIT
```

Comment:
I'm not sure whether it's right to skip `doCommitWithRetry`. It changes the status of the transaction to `invalid`, removes the prepared statement, and so on.

Reply:
How about move line 544 ~ 567 to function `doCommitWithRetry`?